### PR TITLE
SEO Preview: Discern post vs. site preview

### PR DESCRIPTION
**Pull request by dmsnell**
Previously, every preview in the SEO preview pane was showing both a
site and a post preview because it was unaware if we were viewing from
somewhere like **My Sites** instead of from within the **Post Editor**.

Now, the preview component detects if the post editor is currently open
and if it is will fill in the `post` prop to the `<SeoPreviewPane />`
inside of `connect()`

*Note* that this does not fill in the appropriate data _from_ the post
or sites into the preview components; it merely makes the information
and distinction available.

**Testing**
There are two ways of showing the SEO preview (three maybe until we decide for sure whether or not it should show on Theme previews). The **My Sites** site preview and the **Post Editor** post preview. In **master** these previews should all be identical, but afterwards, the **My Sites** preview should show _site_ previews while the **post editor** preview should show post-specific previews.

**Site Preview**
This is small and has the image on the left:
screen shot 2016-08-03 at 3 33 47 pm

**Post Preview**
This is bigger and has the image at the top:
screen shot 2016-08-03 at 3 34 08 pm

cc: @roundhill 

Test live: https://calypso.live/?branch=update/seo-preview-post-data

### CODE REVIEW

**Comment by dmsnell** [comment]
@ehg I feel like I trashed the ignore list here to make this work, but I was stuck between a rock and a hard place. Any advice you have would be incredibly treasured!

**Comment by roundhill** [comment]
Tested and confirmed that the correct `post` object was coming in. LGTM 👍 

Not sure about the ignore list stuff though, will leave that to the experts!

File: server/pragma-checker/index.js [hunk 1]
```diff
@@ -19,7 +19,12 @@ var IGNORED_MODULES = [
 	'lib/upgrades/actions', // nooped on the server as it still uses the singleton Flux architecture
 	'i18n-calypso', // ignore this until we make it work properly on the server
 	'my-sites/themes/thanks-modal', // stubbed on the server until we develop an isomorphic version
-	'my-sites/themes/themes-site-selector-modal' // stubbed on the server until we develop an isomorphic version
+	'my-sites/themes/themes-site-selector-modal', // stubbed on the server until we develop an isomorphic version
+	'components/web-preview', // stubbed on the server until all the dependencies are @ssr-ready
+	'components/seo-preview-pane', // stubbed on the server until all the dependencies are @ssr-ready
+	'reader/stream/post', // stubbed on the server until all the dependencies are @ssr-ready
```
Comment (ehg) [inline server/pragma-checker/index.js]: I think we could take this ignore out?

File: webpack.config.node.js [hunk 2]
```diff
@@ -87,7 +87,9 @@ module.exports = {
 		new webpack.NormalModuleReplacementPlugin( /^lib\/upgrades\/actions$/, 'lodash/noop' ), // Uses Flux dispatcher
 		new webpack.NormalModuleReplacementPlugin( /^lib\/route$/, 'lodash/noop' ), // Depends too much on page.js
 		new webpack.NormalModuleReplacementPlugin( /^my-sites\/themes\/thanks-modal$/, 'components/empty-component' ), // Depends on BOM
-		new webpack.NormalModuleReplacementPlugin( /^my-sites\/themes\/themes-site-selector-modal$/, 'components/empty-component' ) // Depends on BOM
+		new webpack.NormalModuleReplacementPlugin( /^my-sites\/themes\/themes-site-selector-modal$/, 'components/empty-component' ), // Depends on BOM
+		new webpack.NormalModuleReplacementPlugin( /^components\/web-preview$/, 'components/empty-component' ),
+		new webpack.NormalModuleReplacementPlugin( /^components\/seo-preview-pane$/, 'components/empty-component' )
```
Comment (ehg) [inline webpack.config.node.js]: Do we need to NMR this if we've already NMRed `web-preview`?

**Comment by ehg** [comment]
> @ehg I feel like I trashed the ignore list here to make this work, but I was stuck between a rock and a hard place. Any advice you have would be incredibly treasured!

It's a tricky one to navigate. I feel your pain. It's a bit sad that we're ignoring selectors, but seeing as they depend on lots of formatting related modules, I think it's better to ignore them for now — as that's a lot of code to ensure being SSR-ready. Maybe we could have a separate issue for those, as I do see deps like `formatting` and the post normalising utils being useful on the server at some point.

/cc @seear @ockham for thoughts.

BTW, I think `seo-preview-pane` belongs in [`client/blocks`](https://github.com/Automattic/wp-calypso/tree/master/client/blocks#blocks), seeing as it depends on a lot of Calypso.

File: server/pragma-checker/index.js (hunk 1 above)
Comment (dmsnell) [inline server/pragma-checker/index.js]: I reduced the ignore list to what appeared to be a minimal set. Thanks @ehg

File: webpack.config.node.js (hunk 2 above)
Comment (dmsnell) [inline webpack.config.node.js]: Reworked in a rebase to minimize changes here.

**Comment by seear** [comment]
The comments for the new entries in the ignore list state:
`// stubbed on the server until all the dependencies are @ssr-ready`
but this is not quite true. They are being ignored for the ssr pragma check, but still being pulled in on the server build.

To actually stub them, we need to add a couple of lines to the `plugins` array in `webpack.config.node.js`:
```
		new webpack.NormalModuleReplacementPlugin( /^state\/ui\/editor\/selectors$/, 'lodash/noop' ), // will never be called server-side
		new webpack.NormalModuleReplacementPlugin( /^state\/posts\/selectors$/, 'lodash/noop' ), // will never be called server-side
```

This will be safer, because if something changes in the future that means the selectors do get called on the server, it will be a noop instead of something that could cause a problem for SSR.

**Comment by seear** [comment]
...and what is missing framework-wise, which we should try and address soon, is that anything that is stubbed in `webpack.cofig.node.js` should be automatically ignored by the pragma checker.

**Comment by dmsnell** [comment]
> To actually stub them, we need to add a couple of lines to the plugins array in webpack.config.node.js

Thanks @seear. I originally had all of `state` in the webpack config I think but took it out to try and mess with fewer things. Honestly I don't understand it all too well at this point. I have added those provided lines in 9d206c7

**Comment by dmsnell** [comment]
> BTW, I think seo-preview-pane belongs in client/blocks, seeing as it depends on a lot of Calypso.

Sounds good @ehg but I don't think I'll address this here. We have other PRs based on this and I'm afraid it might be more work to adjust now with merge conflicts than it would be to come around after this is all in and refactor.

**Comment by seear** [comment]
>  I have added those provided lines in 9d206c7

Looks good 👍 

> Honestly I don't understand it all too well at this point.

Yes, this is becoming a problem as the `/** @ssr-ready **/` pragma is added to more files. We don't mention stubbing/ignoring in the [docs](https://github.com/Automattic/wp-calypso/blob/master/docs/server-side-rendering.md), and it is not a straightforward topic to cover. 

I'm away for a couple of weeks, but when I get back I'll take another look at how we can check for server-side-rendering compatibility without hindering development of unrelated features.

**Pull request merged**